### ISSUE TITLE
findAsync may be scanning for changes too early

### ISSUE THREAD

**Issue by 20kdc** [Cumcord/Cumcord#8]
findAsync ( https://github.com/Cumcord/Cumcord/blob/stable/src/api/modules/findAsync.js ) detects changes to `webpackChunkdiscord_app`.
However, `webpackChunkdiscord_app` only contains the 'base functions' - actual scanning occurs on `wpRequire.c` (see: https://github.com/Cumcord/Cumcord/blob/stable/src/api/modules/webpackModules.js ).

I've been having some odd behaviour trying to get `SelectImageModal` (avatar selection dialog), where Discord consistently manages to get at SelectImageModal before I can patch it, and I think this behaviour can be explained as such - the `findAsync` check triggers before SelectImageModal actually exists, so it fails, and then triggers for whatever reason after SelectImageModal has been instanced, which is too late for the hook.

For further details, please see https://gitlab.com/20kdc/kdc-cord-plugins/-/blob/main/gcat/core.tsx

**Comment by yellowsink** [comment]
we patch all of the modules in `wpRequire.m` to resolve our callbacks (which will then grab modules from `wpRequire.c` - this has been extensively discussed and tested.

If you intend to patch a lazy loaded module I *highly* recommend you use `@cumcord/patcher/findAndPatch` instead of `findAsync` - we try to have as little as possible using findAsync directly.

PS: I am intrigued by your code style in your linked plugin. I am all for diverse code styles and all, but a couple things that jumped out to me:
 - You call findByProps with `.apply(this, props)`
 - You make an async function and simply return await
 - tsx files with no react in them
 - ~~module.tsx sure is interesting~~

PPS: We don't use Github for pull requests, if you have any changes you wish to propose please refer to https://lists.sr.ht/~creatable/cumcord-devel

**Comment by yellowsink** [comment]
TL;DR
 - there is no "too early" for findAsync to hook
 - if possible dont use findAsync directly

**Comment by yellowsink** [comment]
- functions 
- fair 
- fair 
- spread operator.

**Comment by 20kdc** [comment]
um, what do I use instead of findAsync? findAndPatch isn't applicable due to the warming up vs. powering up barrier (important for, again, fault isolation)

**Comment by yellowsink** [comment]
Make an array of patches and push to that or something. 

Like most plugins do it.

**Comment by 20kdc** [comment]
findAndPatch indeed worked - the question I'm now looking into is "why", but that's not relevant to this issue. Closing.

**Comment by 20kdc** [comment]
Actually, there is something I'd like to note as a sort of post-mortem to this issue:
```
Once a [Promise](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise) is fulfilled or rejected, the respective handler function (onFulfilled or onRejected) will be called asynchronously (scheduled in the current thread loop).
```
 - https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/then

Basically, async can't be trusted to run stuff "as soon as possible".
I'm not sure how findAndPatch gets away with it - maybe it's just a matter of scheduling the exact right amount of promises for things to work while not scheduling enough for things to go too wrong.

**Comment by yellowsink** [comment]
promises always defer things onto the next event loop tick, be it callbacks in `new Promise(...)` or `then`, `catch`, `finally`. That's just how they work.

findAndPatch does absolutely 0 black magic, just clever use of promises and scopes.
[See here](https://git.sr.ht/~creatable/Cumcord/tree/master/item/src/api/patcher/findAndPatch.js]

**Comment by yellowsink** [comment]
Nothing stops us from making it apply synchronously if the module exists (simply two lines), but its really not worth it when findAsync already handles that case for us.

**Comment by 20kdc** [comment]
The problem here is:
+ `SelectImageModal` isn't loaded until it's needed
+ It *will* be loaded when the user interacts with the element
+ Therefore it's a race to hit the "sweet spot" where Discord has loaded `SelectImageModal`, but before Discord's own promise chains return and actually *instance* it.

**Comment by yellowsink** [comment]
Ah, no this is an unsolvable problem with any currently known methods.

If you wish to dig deeper into the inner workings of findAsync go ahead, but without modifying wpRequire and its contents in grotesque ways we don't really think it can be done.

Even old style patching with openContextMenuLazy etc didnt do that

**Comment by yellowsink** [comment]
(though i have just an idea, but i wont discuss it here, ill talk abt it in discord if it works or is promising)

**Comment by Cr3atable** [comment]
Just to chime in here for a second, the thing about Discord modding is that most mods inject after Discord has loaded. There's one exception that loads before and hooks into webpack, but due to Cumcord's scope, we can't do that. The typical solution to these sorts of issues would essentially be to either find another component to hook into and go down the React tree when that component is loaded.

**Comment by 20kdc** [comment]
Don't worry - there's a reason I closed the issue. I have a workaround planned on my end and it will be in place shortly.

**Comment by yellowsink** [comment]
A workaround, indeed please [join](https://discord.gg/FhHQQrVs7U) the discord, I am legitimately intrigued by how you plan to solve this

**Comment by 20kdc** [comment]
Said workaround is in place now and has been pushed to CI on my end (it's the one I posted to Discord as asked)
